### PR TITLE
perf(rust, python): SIMD accelerated `arg_min`/`arg_max` (via `argminmax`)

### PR DESCRIPTION
**Pull request by jvdd**
This PR uses [`argminmax`](https://github.com/jvdd/argminmax) to provide faster `arg_min` and `arg_max` operations.

## Motivation

After playing around with writing a custom trait for `argmin` & `argmax` I observed considerable speed ups when using the argminmax crate over polars. Eventually, I ended up diving in the polars codebase and integrated argminmax - which resulted in this PR.

I conducted a *very* basic benchmark and found:
- a ~4.5x improvement for a single chunk (internally handled as a slice)
- a ~25x improvement for multiple chunks (6 chunks of 10k i32 values)

The benchmark code + results are included at the bottom.

## Changes made
*(I'll discuss the changes for `arg_min`, those for `arg_max` are identical)*

Minor changes were made to the previous code, where the `.iter().reduce()` implementation is now only used for `Utf8Chunked`. Since this implementation is only used for string data, I renamed the `arg_min` to `arg_min_str`.

Updated code (using argminmax):
- **slice**: `arg_min_slice` now calls `vals.argmin()`
    - renamed this to `arg_min_numeric_slice` (since it is only called on numeric dtypes)
- **chunked array**:
    - calls argminmax on chunks (does not even require the "arrow2" feature as `.values().as_slice()` is called on the chunks)
    - renamed `arg_min` -> `arg_min_numeric` (since it is only called on numeric dtypes)

A big red flag in the current code is that `argminmax` assumes non-empty slices, resulting in adding `.is_empty` checks for each chunk or slice. :thinking: 

## TODOs
- [x] create a new release of argminmax (that contains the now optimized argmin & argmax functions)
- [ ] add more extensive testing for various edge cases (null chunks, multiple chunks with nulls, nan-handling, ...)

## Potential improvements / considerations
I think the following three points are worth considering
- [ ] I could adapt argminmax to return Option instead of usize (as such making it return None for empty arrays)
- [x] We should make sure that floats are handled identical - I believe currently `polars` ignore nans for `arg_min` and `arg_max` which complies with how the `ArgMinMax` trait handles this.
  - [x] is it correct that `float16` is not supported by `polars`?
- [ ] It might be worth considering adding .arg_min_max to get both in one pass over the data. As this task is mainly bound by CPU memory bandwidth, it is unfortunate that the data needs to be passed twice to the CPU when interested in both argmin and argmax values. The benchmark results below demonstrate that `argminmax` can extract both values in approximately the same time required to extract either one.

---

### Code

`main.rs` :arrow_down:
```rs
use argminmax::ArgMinMax;
use polars::prelude::*;
use rand::prelude::*;

// import black_box to prevent the compiler from optimizing away the loop
use std::hint::black_box;

// ----------------------------------------------------------------
// basic benchmark
// ----------------------------------------------------------------

fn random_i32_vec(n: usize) -> Vec<i32> {
    let mut rng = rand::thread_rng();
    (0..n).map(|_| rng.gen_range(-100..100)).collect()
}

fn main() {
    // Create an empty polars series
    let s = Series::new("a", Vec::<i32>::new());
    assert_eq!(s.chunks().len(), 1);
    assert_eq!(s.arg_min(), None);
    assert_eq!(s.arg_max(), None);
    // Multiple empty chunks
    let mut s = Series::new("a", Vec::<i32>::new());
    _ = s.append(&Series::new("a", vec![1i32, 2i32]));
    _ = s.append(&Series::new("a", Vec::<i32>::new()));
    _ = s.append(&Series::new("a", vec![1i32, 2i32]));
    _ = s.append(&Series::new("a", Vec::<i32>::new()));
    assert!(s.chunks().len() > 2);
    assert_eq!(s.arg_min(), Some(0));
    assert_eq!(s.arg_max(), Some(1));

    println!("Benchmarking single chunk of 10k elements");

    // Create a polars i32 series of 10k elements (1 chunk)
    let mut v = random_i32_vec(10_000);
    v[376] = 1000;
    v[7853] = -1000;
    let s = Series::new("a", v.clone());
    assert_eq!(s.chunks().len(), 1);

    // polars
    let start = std::time::Instant::now();
    for _ in 0..10 {
        let max_idx = black_box(s.arg_max().unwrap());
        assert_eq!(max_idx, 376);
        let min_idx = black_box(s.arg_min().unwrap());
        assert_eq!(min_idx, 7853);
    }
    let end = std::time::Instant::now();
    println!("polars: {:?}", end - start);

    // argminmax
    let start = std::time::Instant::now();
    for _ in 0..10 {
        let argminmax_idx = black_box(v.argmax());
        assert_eq!(argminmax_idx, 376);
        let argminmax_idx = black_box(v.argmin());
        assert_eq!(argminmax_idx, 7853);
    }
    let end = std::time::Instant::now();
    println!("argminmax: {:?}", end - start);

    let start = std::time::Instant::now();
    for _ in 0..10 {
        let (min_idx, max_idx) = black_box(v.argminmax());
        assert_eq!(min_idx, 7853);
        assert_eq!(max_idx, 376);
    }
    let end = std::time::Instant::now();
    println!("argminmax both in 1 pass: {:?}", end - start);

    println!("");
    println!("Benchmarking chunked series of 6 chunks of 10k elements each");

    // Create a polars i32 series that consists of 6 chunks
    let mut v = random_i32_vec(10_000);
    v[376] = 1000;
    v[7853] = -1000;

    let mut s = Series::new("a", v.clone());
    // append 5 chunks
    for _ in 0..5 {
        _ = s.append(&Series::new("a", v.clone()));
    }
    assert_eq!(s.chunks().len(), 6);

    // polars
    let start = std::time::Instant::now();
    for _ in 0..10 {
        let max_idx = black_box(s.arg_max().unwrap());
        assert_eq!(max_idx, 376);
        let min_idx = black_box(s.arg_min().unwrap());
        assert_eq!(min_idx, 7853);
    }
    let end = std::time::Instant::now();
    println!("polars: {:?}", end - start);
}
```

`cargo.toml` :arrow_down:

```toml
[package]
name = "polars_argmm"
version = "0.1.0"
edition = "2021"

[dependencies]
polars = { version = "0.28.0", features = ["simd"] }
# comment the above line and uncomment the below line to bench this PR
# polars = { git = "https://github.com/jvdd/polars", branch = "argminmax_integration" }
argminmax ={ git = "https://github.com/jvdd/argminmax" }
rand = "0.8.5"
rand_distr = "0.4.3"
```

### Results

`polars = { git = "https://github.com/jvdd/polars", branch = "argminmax_integration" }` :arrow_down: 
```
Benchmarking single chunk of 10k elements
polars: 63.706µs
argminmax: 61.45µs
argminmax both in 1 pass: 32.693µs

Benchmarking chunked series of 6 chunks of 10k elements each
polars: 375.618µs
```

`polars = { version = "0.28.0", features = ["simd"] }` :arrow_down: 
```
Benchmarking single chunk of 10k elements
polars: 295.682µs
argminmax: 85.043µs
argminmax both in 1 pass: 41.74µs

Benchmarking chunked series of 6 chunks of 10k elements each
polars: 9.817482ms
```

### CODE REVIEW

**Comment by ritchie46** [comment]
Thanks @jvdd that are some impressive benchmarks. :raised_hands: 

I have left some comments.

**Comment by jvdd** [comment]
Thank you @ritchie46! It seems as though there are no comments over here 😅

**Comment by ritchie46** [comment]
> Thank you @ritchie46! It seems as though there are no comments over here sweat_smile

Whoops. I forgot the hit the `finish review` botton! :see_no_evil:

File: polars/polars-ops/Cargo.toml [hunk 1]
```diff
@@ -10,6 +10,7 @@ description = "More operations on polars data structures"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
+argminmax = { git = "https://github.com/jvdd/argminmax" }
```
[the remark this answers is not in the document]
Comment (jvdd) [reply polars/polars-ops/Cargo.toml:13]: I'll create a new release shortly! Just want to double check whether you are fine (with currently) keeping the assumption of no empty array..

[the remark this answers is not in the document]
Comment (jvdd) [reply polars/polars-ops/Cargo.toml:13]: I'll create a new release shortly! Just want to double check whether you are fine (with currently) keeping the assumption of no empty array..

File: polars/polars-ops/src/series/ops/arg_min_max.rs [hunk 2]
```diff
@@ -222,32 +217,108 @@ where
...
+        IsSorted::Ascending => Some(ca.len() - 1),
+        IsSorted::Descending => Some(0),
+        IsSorted::Not => {
+            ca.downcast_iter()
+                .fold((None, None, 0), |acc, arr| {
+                    if arr.len() == 0 {
+                        // argminmax assumes not empty
+                        return acc;
+                    }
+                    let chunk_max_idx: usize = arr.values().as_slice().argmax();
```
Comment (ritchie46) [inline polars/polars-ops/src/series/ops/arg_min_max.rs:277]: This seems to ignore the null values that are set in the validity bits. The values that are masked out can have any arbitrary value, so we are not allowed to assume anything here.

I think that if `null_count == 0`, we can continue with this fast argmax branch, but if not we should take the old branch to ensure correctness.

Comment (jvdd) [reply polars/polars-ops/src/series/ops/arg_min_max.rs:277]: I'm not entirely sure about this, but I believe that the current polars code also doesn't consider the validity bits (at least in the case of a single chunk) - see code below.

However, I'm certainly open to add this functionality to argminmax https://github.com/jvdd/argminmax/issues/14#issuecomment-1495658702
Just wondering about what the desired / default behavior would be; ignoring the NULL indices?

---

Code :arrow_down: 
```rs
    let s = Series::new("a", vec![Some(1), None, Some(3)]);
    let arr = s.i32().unwrap().chunks()[0].as_ref();
    println!("{:?}", arr);
    println!("{:?}", arr.validity().unwrap());
    println!("argmin: {:?}", s.arg_min().unwrap()); // returns 1 - which is None...
    println!("argmax: {:?}", s.arg_max().unwrap());
```

Output :arrow_down: 
```
Int32[1, None, 3]
[0b_____101]
argmin: 1
argmax: 2
```

Comment (ritchie46) [reply polars/polars-ops/src/series/ops/arg_min_max.rs:277]: That's because `None < Some(value)` in our ordering. I think for now we should only add this for cases where all values are valid. Adding an arrow dependency via `jvdd/argminmax` will definitely run out of sync meaning we compile arrow twice, which is a huge dependency.

Let's start with the easy case first and then we can figure out the more complicate (nulls) in a later time. :)

Comment (jvdd) [reply polars/polars-ops/src/series/ops/arg_min_max.rs:277]: You are completely right. Did not realize there was an ordering for optional values. I'll update this PR accordingly!

Furthermore - I do think it should be possible to integrate this into argminmax, without necessarily depending on `arrow2`! If you can pass smth like an `Option<BitMask>` to the argminmax functions, you can achieve this functionality without having to depend on `arrow2` :thinking:

File: polars/polars-ops/src/series/ops/arg_min_max.rs [hunk 3]
```diff
@@ -222,32 +218,137 @@ where
...
+                    let chunk_min_idx: Option<usize>;
+                    let chunk_min_val: Option<T::Native>;
+                    if arr.null_count() > 0 {
+                        // When there are nulls, we should compare Option<T::Native>
+                        chunk_min_val = None; // because None < Some(_)
+                        chunk_min_idx = arr
+                            .into_iter()
+                            .enumerate()
+                            .reduce(|acc, (idx, val)| if acc.1 > val { (idx, val) } else { acc })
+                            .map(|tpl| tpl.0);
```
Comment (jvdd) [inline polars/polars-ops/src/series/ops/arg_min_max.rs:244]: Basically we should return the index of the first null value

**Comment by jvdd** [comment]
If you can point me to where add some more extensive tests (either in Python or Rust) - I’ll gladly include some in this PR :)

**Comment by jvdd** [comment]
I've added the `"nightly_simd"` as a (default) feature to argminmax since most `x86`/`x86_64` SIMD intrinsics have been stabilized in Rust (except for `AVX512`). Leaving this compute power (`SSE` & `AVX(2)` unused would have been a pity :wink: 

With this feature, the unstable  `AVX512` and `NEON` SIMD are toggled. I've created a new release that includes this feature and updated the code to disable it by default. When the user enables the polars `"simd"` feature, this `"nightly_simd"` feature will be enabled.

**Comment by ritchie46** [comment]
> With this feature, the unstable AVX512 and NEON SIMD are toggled. I've created a new release that includes this feature and updated the code to disable it by default. When the user enables the polars "simd" feature, this "nightly_simd" feature will be enabled.

That's great. And on `stable`? It will default to a fallback?

**Comment by jvdd** [comment]
> > With this feature, the unstable AVX512 and NEON SIMD are toggled. I've created a new release that includes this feature and updated the code to disable it by default. When the user enables the polars "simd" feature, this "nightly_simd" feature will be enabled.
> 
> That's great. And on `stable`? It will default to a fallback?

Yes!
